### ISSUE TITLE
Need to improve time required for initial `git clone` after startup of Che 7 workspace on osio

### ISSUE THREAD

**Issue by ibuziuk** [redhat-developer/rh-che#1012]
gluster storage FS access [1] is slow on osio and `cloning projects` always takes sufficient amount of time. In order to decrease the time required for cloning project after workspace startup we need to investigate possibility of using some flags e.g. `--depth` [2]  initial cloning of the project used for in Che 7. 

[1] https://docs.google.com/spreadsheets/d/1P9B90RXc2BnicR9qfNTeKuQ--8m4K7eLYixgkW_d6t4/edit#gid=0
[2] https://git-scm.com/docs/git-clone#git-clone---depthltdepthgt

**Comment by ibuziuk** [comment]
@l0rd AFAIK, currently in che 7 / Theia  on osio is possible to do `git clone` only via terminal, e.g. even if project is added during workspace creation via dashboard it will not be cloned automatically. So, I am a bit baffled what exactly we should investigate since currently Che 7 / Theia does not support auto-clone mechanism similar to what Che 6 has. Could you please clarify ?

**Comment by l0rd** [comment]
The idea here is to investigate on `git clone` options to make cloning as fast as possible and to address problems like https://github.com/redhat-developer/rh-che/issues/772. This will have an impact on Che 6 now, as well as on Che 7 when it will be possible to specify a project for those stacks too (very soon).

**Comment by ibuziuk** [comment]
sure, we would investigate git clone options, but in order to provide the real fix to Che 7 we need to know how exactly git clone API works in Theia to provide support of some flag e.g. `--depth`. Or this issue is only about investigation and implementation of github options would be done by another team ?

**Comment by l0rd** [comment]
I think the work on GitHub cloning for Theia is already in progress (@sunix
do you know who is working on this?).

This issue should be investigation + implementation for Che 6 + share your
findings with who is working on che 7.

What already exists on che 7 is the 'git clone' command in the command
palette. Part of this task could be adding a new  'git clone swallow' to
Che 7 command palette.

But really, whatever option you find, implementation will be trivial. You
need to focus more on the investigation part: find the options that make
git cloning as fast as possible.

Il mer 24 ott 2018, 20:32 Ilya Buziuk <notifications@github.com> ha scritto:

**Comment by l0rd** [comment]
@ibuziuk any update on this one? cc @gorkem

**Comment by ibuziuk** [comment]
@l0rd it looks like the fastest possible way of shallow cloning is using the following syntax: `git clone -b <branch> --single-branch --depth 1 <repository>`. However, this approach does not really help when cloning inside of gluster volume. Here are results cloning `che` repo: 

1. vanilla git clone in PVC ~`6m48s`

```
[user@workspace6tmtqy5o6xvftg38 projects]$ time git clone https://github.com/eclipse/che.git
Cloning into 'che'...
remote: Enumerating objects: 304, done.
remote: Counting objects: 100% (304/304), done.
remote: Compressing objects: 100% (186/186), done.
remote: Total 444533 (delta 208), reused 128 (delta 103), pack-reused 444229
Receiving objects: 100% (444533/444533), 96.90 MiB | 1.97 MiB/s, done.
Resolving deltas: 100% (194916/194916), done.
Checking out files: 100% (12698/12698), done.

real    6m48.819s
user    0m15.710s
sys     0m19.846s  
```

2. shallow git clone in PVC  ~ `5m31s`

```
[user@workspace6tmtqy5o6xvftg38 projects]$ time git clone -b 'master' --single-branch --depth 1 https://github.com/eclipse/che
Cloning into 'che'...
remote: Enumerating objects: 20671, done.
remote: Counting objects: 100% (20671/20671), done.
remote: Compressing objects: 100% (12850/12850), done.
remote: Total 20671 (delta 4656), reused 12145 (delta 2759), pack-reused 0
Receiving objects: 100% (20671/20671), 13.10 MiB | 3.58 MiB/s, done.
Resolving deltas: 100% (4656/4656), done.
Checking out files: 100% (12698/12698), done.

real    5m31.733s
user    0m2.083s
sys     0m6.298s

```
3.vanilla `git clone` in emptyDir volume `~11s`

```
time git clone  https://github.com/eclipse/che
Cloning into 'che'...
remote: Enumerating objects: 304, done.
remote: Counting objects: 100% (304/304), done.
remote: Compressing objects: 100% (186/186), done.
remote: Total 444533 (delta 208), reused 128 (delta 103), pack-reused 444229
Receiving objects: 100% (444533/444533), 96.90 MiB | 25.74 MiB/s, done.
Resolving deltas: 100% (194916/194916), done.

real    0m10.945s
user    0m10.103s
sys     0m4.176s

```
4. shallow `git clone` in emptyDir volume ~ `7s`

```
[user@workspace6tmtqy5o6xvftg38 projects]$ time git clone -b 'master' --single-branch --depth 1 https://github.com/eclipse/che
Cloning into 'che'...
remote: Enumerating objects: 20671, done.
remote: Counting objects: 100% (20671/20671), done.
remote: Compressing objects: 100% (12850/12850), done.
remote: Total 20671 (delta 4656), reused 12145 (delta 2759), pack-reused 0
Receiving objects: 100% (20671/20671), 13.10 MiB | 12.39 MiB/s, done.
Resolving deltas: 100% (4656/4656), done.

real    0m7.319s
user    0m1.226s
sys     0m1.687s
```

5. vanilla `git clone` in container FS 

 
6. shallow `git clone` in container FS

**Comment by ibuziuk** [comment]
The funny thing is that when cloning in gluster volume there is a `Checking out files` phase that takes `99%` of the `git clone` time and results in more than 5 minutes for shallow cloning of the . This phase does not seem to be presented / logged when cloning in container FS or `emptyDir` volume, where clone time takes just a couple of seconds. 

@l0rd what I wanted to try is updating git version used in stacks (currently `1.8.3.1` is used) and time cloning against newer version of git. Other than that, currently I do not have good ideas about how clone operation could potentially be improved from the `che` side. Maybe it would be relevant to get some input from SD / storage team regarding this situation with utterly slow cloning ?

Also, I tried obtaining source code without `git` and using `curl` and unzipping the archive instead e.g.

> curl -LO https://github.com/eclipse/che/archive/master.zip
> unzip master.zip

However, it also resulted in sufficient amount of time ~ 9m (in container FS it takes a couple of seconds)

**Comment by gorkem** [comment]
Another idea is to clone into container FS, work from there and sync it to `/projects` once `/projects` is equal, we can update the IDE to point to `/projects` on the attached PV.

**Comment by ibuziuk** [comment]
I tried to clone che repo in container FS and than copy it to `/projects` and this took more than 5 minutes:

```
[user@workspace866w2m68unsivdxi tmp]$ time cp -r che /projects/

real    5m17.324s
user    0m0.258s
sys     0m5.474s
```

Also, I tried updating git to more recent version (`2.18.0`) and perform the same test with cloning, but this does not seem to change anything:

```
[user@workspace866w2m68unsivdxi projects]$ time git clone -b 'master' --single-branch --depth 1 https://github.com/eclipse/che
Cloning into 'che'...
remote: Enumerating objects: 20672, done.
remote: Counting objects: 100% (20672/20672), done.
remote: Compressing objects: 100% (12855/12855), done.
remote: Total 20672 (delta 4656), reused 12145 (delta 2755), pack-reused 0
Receiving objects: 100% (20672/20672), 13.10 MiB | 3.50 MiB/s, done.
Resolving deltas: 100% (4656/4656), done.
Checking out files: 100% (12699/12699), done.

real    6m21.118s
user    0m2.192s
sys     0m7.034s
```

The image with updated git version is `ibuziuk/centos_jdk8:git2`

@gorkem @l0rd I believe we can close this issue since it looks like it is not possible to sufficiently improve cloning time in the gluster volume using shallow cloning. We could  possibly open another issue for investigating possibility of starting ephemeral workspace + sidecar pod with mounted PVC and using rsync, but it is very likely that we would still hit the problem with utterly slow sync process. WDYT ?

**Comment by garagatyi** [comment]
@ibuziuk I believe that sync approach won't be that slow. The problem is that gluster doesn't like a lot of operations with files. But the overall speed when a lot of operations are not involved is not that bad.
Since sync operation is dependent on every file change it can sync every X seconds and gluster would definitely tolerate this.
What are the implications of that approach:
- workspace start is fast even without the gluster subvol fix
- workspace stop is not that fast anymore since we would need to do shutdown sync. But this might not be very problematic. 
- We would need to tune workspace stop flow to stop sidecar after stop of other containers
- We can make sync sidecar a separate deployment to stop it separately from the IDE. Or we can use a container in a pod where another /projects requiring container runs. In that case, sync sidecar would be able to sync a shared ephemeral volume and sync everything locally without involving network!
- If we use separate sync sidecar we would also be able to implement other sync features, such as let containers run in different pods and sync some folders between them. But this is out of the scope of this issue. 
- we could even start that sidecar before the Che 7 plugin brokering process and let brokers sync plugins with the deployment of a workspace that will start later. This would allow us to have completely ephemeral workspaces without the need to run brokers twice as we do for ephemeral workspaces now (PR is not merged yet)
- this would also allow us to restart deployments of user's prod app without restarting tooling (e.g. when a user wants to deploy his app compiled in maven container and docker image built witt buildah) 

Might seem like a list of crazy ideas, but it doesn't seem that we could proceed by expecting improvements from the gluster side.

**Comment by amisevsk** [comment]
To complete the discussion on moving files in and out, I was able to copy the Che source tree *out* of the PVC to the container filesystem in around 1m30s

**Comment by l0rd** [comment]
@garagatyi I think having a persistence-plugin (that basically rsync /projects/) is the way to go for the future. But I also think that this can be a tricky task and I don't want us to be in a hurry when implementing that. And unfortunately we are in a hurry: Che 7 workspaces on osio can take more than 2 minutes to start and this makes them unusable.

So what should we do in the short term? Improving performances for ephemeral workspaces! The goal should be to have ephemeral workspaces started in under 5 seconds on OSIO. Really think about it: ephemeral is perfectly fine for a free service, nobody would complain about it because they are not paying, our users would be delighted with a fast, extensible free IDE in the cloud. In parallel we can start thinking about a persistence plugin that would be useful when we have a terrible storage like gluster but that should not be our n.1 priority, we have time to find the best solution.

I discussed that with @ibuziuk before. Even if shallow cloning and local registry didn't provided the expected results we still have a couple of options that could dramatically improve ephemeral workspaces bootstrapping:

1. **Images pulling**: @amisevsk [analysis](https://github.com/redhat-developer/rh-che/issues/1014) shows that this is by far our bottleneck. The good news is that pulling an image is, sometimes, really fast. If we understand why pulling is fast we may be able to make it always fast! We need to continue looking at it: it's critical.
2. **Load plain theia first**: we currently start Theia after all brokers have been pulled and run, all plugins have been pulled and container created. We need to make it faster. We need to have Theia loaded in the user browser first and, in a second step, start all the rest. Something that comes to my mind to achieve this is to create a first k8s deployment with only one pod/container/service (theia) and, after we have started, start a rolling update of it with a new deployment definition that has all the pod/containers/services for the plugins. That's just an idea and I am sure there are other way to achieve that but the goal should be clear: starting a workspace should not take more than starting standalone Theia (5s or so).

Point 1. is specific to OSIO, 2. is not. It's an upstream improvement and would dramatically improve Che UX.

Another thing that we may still look at is reducing images size. This can be achieved using distroless images (or `FROM scratch`)  for brokers, editors and plugins. This has lower priority because if we fix 1. and 2. we would not download the images on OSIO. But anyway its' a quick win, could be a panacea until 1. and 2. get addressed, and would have a great benefit for local Che bootstrap.

**Comment by garagatyi** [comment]
@l0rd I personally think that ephemeral workspaces would provide a poor experience to users:
- the user loses his un-pushed changes on workspace restart and workspace is a subject to timeout when the user went to grab a coffee
- the user loses downloaded dependencies, so building non-dummy maven or node project would involve downloading all the dependencies each time

Yes, start time with ephemeral users is a win, but UX is not that good.
And if we think that it would be OK for a user to experience that with a free service then users might use another public IDE service that does provide persistence.

**Comment by l0rd** [comment]
@garagatyi we need to make che startup faster. We can discuss a lot about if is ephemeral acceptable or not. But that doesn't change much our priorities: even if I think `rsync` is important it won't help much making the startup faster. Image pulling and loading the editor before everything else should be our priorities.

**Comment by amisevsk** [comment]
Is it possible to uncouple the Theia browser loading from workspace start? I can scale a deployment to zero and the in-browser editor just shows the yellow bar across the bottom. If we can start both in parallel, it might help with startup time.

Regarding image pulling -- I have no idea what is going on when the pull takes no time, as it seems to be violating the always pull admission controller (not that I'm complaining...)

**Comment by amisevsk** [comment]
To add to the Che 7 part: Without image pull times, starting the back end components *could* be done in around 15 seconds without changing anything -- the charts [here](https://github.com/redhat-developer/rh-che/issues/1014#issuecomment-437043526) show that without the pulls, the only significant action is Theia start. Note also that the pinkish block (plugin broker) should already be much shorter since the majority of the wait there is for the PVC mount, which is fixed by https://github.com/eclipse/che/pull/11657

**Comment by ibuziuk** [comment]
Closing this issue since according to the research it will not be possible to sufficiently improve cloning time via shallow option in gluster volume (still takes ~ 5-6 minutes for che repostitory)
